LPS-104983 Add Network Status in Toolbar

matuzalemsteles opened the pull request.

liferay-continuous-integration commented:
To conserve resources, the PR Tester does not automatically run for every pull.

If your code changes were already tested in another pull, reference that pull in this pull so the test results can be analyzed.

If your pull was never tested, comment &quot;ci:test&quot; to run the PR Tester for this pull.

matuzalemsteles commented on modules/apps/layout/layout-content-page-editor-web/src/main/resources/META-INF/resources/page_editor/core/hooks/useService.js:
```diff
@@ -0,0 +1,122 @@
+	...action
+});
+
+const useService = config => {
```
Moving the implementation of `serviceFetch` here was important to build network logic and easily share this information, I think it turned out to be a more integrated design with the application. So every time someone send to the backend some information the status bar will notify the user or other components can change the state to improve UX.

Maybe in the future we can replace this implementation with `ClayDataProvider` features, soon we will have more features to support these edge cases, allowing to control parallel requests, manage requests when the network goes down, request deduplication among other features we already have.

wincent replied:
Some things to bear in mind when moving anything from a function to a hook is:

- A hook is just a function that can use other hooks.
- So, one good reason to do so is because you want to use other hooks!
- But once you've made something into a hook you can now no longer use it outside of a React component; this may be a large downside, depending on the use case.
- One of the main use cases for hooks, if not *the* main use case, is to facilitate reuse and composition. If your hook is used in only one place, as this one is, then you pay the cost of an additional layer of indirection but you don't reap the benefit of abstraction until it is used in other places (and it may never be).
- If there is any doubt about actual reuse, dumping the body of a hook into the function is exactly equivalent to using the hook, so you can postpone some of the risk of prematurely abstracting. (Example from this project: `useThunk` is used in exactly one place, but it is designed with reuse in mind *and* it is sufficiently tiny that you can actually reason about how it is likely to behave in different scenarios that we have a better shot of getting the abstraction right on the first try).

All of which is to say, we might want something like `useService`, but we should tread carefully.

matuzalemsteles replied:
Of course, the idea of ​​`useService` is not to be like `useResource` in Clay, we are loading the data, here we will be sending it, so I think this is being pretty generic for your role. Since I need to deal with network state and other information.

> One of the main use cases for hooks, if not the main use case, is to facilitate reuse and composition. If your hook is used in only one place, as this one is, then you pay the cost of an additional layer of indirection but you don't reap the benefit of abstraction until it is used in other places (and it may never be).

Although I almost always agree with this view, I also see the hooks with an alternative to `render props` or a `hoc`, since their goal is just to deal with some logic and states and not render "nothing".

> All of which is to say, we might want something like useService, but we should tread carefully.

I'm drawing the conclusion that you are against the `useService` ​​hook haha.

I'll go the other way, keep the `serviceFetch` function and add a new callback to dispatch the changes to the store. This will seem a bit "manual" as we will have to pass `dispatch` to service for all `serviceFetch` calls.

wincent replied:
> Although I almost always agree with this view, I also see the hooks with an alternative to `render props` or a `hoc`, since their goal is just to deal with some logic and states and not render "nothing".

Sure, hooks can replace other patterns, but my point is that when you create a custom hook that is used in only one place, you are adding a layer of indirection that would not be there if you had just copied the contents of your custom hook inline (eg. in the case of `useService`, you have a `useReducer` call, a `useCallback`, and some logic; and you could have inlined that all in the one place it is used without adding the layer of indirection and it would still behave exactly the same). The other obvious thing you get/do when you add that layer of direction is that you name the functionality as opposed to just dumping it inline "in situ". So, when you create the hook, you have something that is (at least thoretically) reusable and abstracted, which brings with it both potential benefits (eg. re-use) and costs (eg. maintaining API surface area) which you don't really get to receive or pay until that second use comes along, if it ever does.

matuzalemsteles commented:
cc @p2kmgcl

wincent commented on modules/apps/layout/layout-content-page-editor-web/src/main/resources/META-INF/resources/page_editor/app/components/NetworkStatusBar.js:
```diff
@@ -0,0 +1,65 @@
+				<span className="lfr-portal-tooltip navbar-text">{status}</span>
+			)}
+		</li>
+	);
```
Do we really want to return an empty `li` when there is no status? Wouldn't it be better to just render nothing in that case?

wincent commented on modules/apps/layout/layout-content-page-editor-web/src/main/resources/META-INF/resources/page_editor/app/index.js:
```diff
@@ -36,13 +37,18 @@ const {useReducer} = React;
-		<ConfigContext.Provider value={config}>
-			<StoreContext.Provider value={store}>
+		<ConfigContext.Provider value={memoizedConfig}>
+			<StoreContext.Provider value={{...store, service}}>
```
I don't like the idea of mixing a `fetcher` function into the `ConfigContext` like this. A function that has a side effect is not "configuration". To keep things simple, the mental model for `ConfigContext` is that is should be static read-only data from the server that we can read, *not* data + functions.

matuzalemsteles replied:
No problem, we can continue to maintain this `ConfigContext` mental model, it was just a way to get to the services, since we are always passing `config` to them I did not want to have to always pass a fetcher, I wanted to hide under-the-table. To be something more "automatic".

I will think of another way. Any idea is valuable.

wincent replied:
There a few alternatives that spring to mind that all have there own tradeoffs. One is a separate context that we use to pass around this kind — "utility"/"service" functions — of thing (naming it would be hard). Another is making the service module into a configurable singleton (which would make it easy to use but with the downsides that always come with using globals/singletons, and with the folks working on the page editor have so far tried pretty hard to avoid). And the least glamorous option of all is to just pass the fetcher around explicitly, which has the nice property that it allows you to defer making a call on how to abstract this and this mitigate the risk of making the wrong abstraction.

julien commented on modules/apps/layout/layout-content-page-editor-web/src/main/resources/META-INF/resources/page_editor/app/components/NetworkStatusBar.js:
```diff
@@ -0,0 +1,65 @@
+	useEventListener('online', () => setIsOnline(true), true, window);
+
+	useEventListener('offline', () => setIsOnline(false), true, window);
+
```
@matuzalemsteles I like the idea but I don't think there's anything defined (in terms of UX) for that use case, so I'd say that until we don't know what we want we shouldn't move that to the store. 

@wincent what are your thoughts on this?

wincent replied:
I wouldn't move it until we have a clear idea for how we would use it. As a general practice, keeping things local means you can change them without breaking other parts of the system; as soon as you move something somewhere central or generalize it, you immediately introduce the possibility of dependency and therefore have to deal with potential breakage. You want to make sure the juice is worth the squeeze.

The other thing I would say is that if we do decide to make this somehow central, we should evaluate various options for doing so. If we cram too much into the `useService` hook then we are potentially increasing the coupling between the hook, the store, and everything that relies directly or indirectly on the hook, which may make it harder to change the system. Ideally we would favor a number of smaller, composable abstractions — at least initially — over ones which accrue functionality (same issue: these easily become bloated, hard to change etc).

Riffing on the old theme: Premature Abstraction Is The Root Of All Evil™.

matuzalemsteles replied:
Yeah, sounds good. We can incrementally increase as needed and refactor if necessary.

julien commented on modules/apps/layout/layout-content-page-editor-web/src/main/resources/META-INF/resources/page_editor/app/index.js:
```diff
@@ -36,13 +37,18 @@ const {useReducer} = React;
 	);
 
+	/* eslint-disable-next-line react-hooks/exhaustive-deps */
+	const memoizedConfig = useMemo(() => ({...config, fetcher}), []);
```
@matuzalemsteles just so you know, the `useEffect` to save `layoutData` in `PageEditor` is soon going to be removed, and `layoutData` will be "updated" on the backend.

matuzalemsteles replied:
Oh thanks for the information

wincent commented on modules/apps/layout/layout-content-page-editor-web/src/main/resources/META-INF/resources/page_editor/core/hooks/useService.js:
```diff
@@ -0,0 +1,122 @@
+	return {fetcher, ...state};
+};
+
+export default useService;
```
Other hooks in this project just do:

```javascript
export default useThing() {
  // ...
}
```

We should do the same here.

wincent commented on modules/apps/layout/layout-content-page-editor-web/src/main/resources/META-INF/resources/page_editor/core/hooks/useService.js:
```diff
@@ -0,0 +1,122 @@
+const initialState = {
+	error: null,
+	lastFetch: null,
+	networkStatus: SERVICE_NETWORK_STATUS_TYPES.Unused
```
I think this could just be `status`.

matuzalemsteles commented on modules/apps/layout/layout-content-page-editor-web/src/main/resources/META-INF/resources/page_editor/app/config/constants/serviceNetworkStatusTypes.js:
```diff
@@ -0,0 +1,25 @@
+ */
+
+/**
+ * - Unused (0) When no request is happening the status will be `unused`.
```
Hm yeah, it's better

matuzalemsteles commented on modules/apps/layout/layout-content-page-editor-web/src/main/resources/META-INF/resources/page_editor/app/config/constants/serviceNetworkStatusTypes.js:
```diff
@@ -0,0 +1,25 @@
+
+/**
+ * - Unused (0) When no request is happening the status will be `unused`.
+ * - Fetching (1) When no request is happening the status will be `unused`.
```
Oh yeah, sorry. I think I ended up copying 😅

matuzalemsteles commented:
hey @julien and @wincent I went ahead and changed the approach, kept the `serviceFetch` function with a `onServiceStatus` method which is the signature to pass to `dispatch` which will save the data in the `Store`. These changes are in the last commit.

These changes are not yet ready to be sent but I would like a different view on this. I also had some issues with `serviceReducer` which ended up creating new store referrals, but this causes an infinite loop due to the layoutDate update in PageEditor, so I had to update only services in the Store, but for some reason the status when changed not being reflected in the `<NetworkStatusBar />` component, I thought it was the internet that was very fast, but then I switched to slow 3G, but it still does not change, the interesting thing is that the service changes but not in the component.

julien commented:
Hey @matuzalemsteles let's try a Network only context to see if we can avoid this infinite loop.
The [`useEffect`](https://github.com/liferay/liferay-portal/blob/fd1371543e5331e6c72931074608140e6e97e72f/modules/apps/layout/layout-content-page-editor-web/src/main/resources/META-INF/resources/page_editor/app/components/PageEditor.js#L287-L302) will eventually disappear, but not right now.

matuzalemsteles commented on modules/apps/layout/layout-content-page-editor-web/src/main/resources/META-INF/resources/page_editor/app/services/index.js:
```diff
@@ -0,0 +1,42 @@
+			{children}
+		</NetworkContext.Provider>
+	);
+};
```
Network information is now in its own context to minimize infinite looping in the Store. One downside to this is that everything here still seems manual, whenever you have to create a thunk you will need to pass the fetcher to the service, which may seem strange but if it wasn't done we would still have to keep doing it to pass the callback of `onServiceStatus` for each thunk.

Another option for this is to go with the idea of ​​the `useService` hook that will dock with `useThunk`, so any thunk will have `fetcher` visibility, making it less painful to call the network context every time you use a thunk.

cc @wincent @julien

matuzalemsteles replied:
I was thinking of following a feature that exists in Redux thunk which is `withExtraArgument`, of course here would be done otherwise.

matuzalemsteles commented on modules/apps/layout/layout-content-page-editor-web/src/main/resources/META-INF/resources/page_editor/app/thunks/addFragment.js:
```diff
@@ -19,18 +19,20 @@ export default function addFragment({
 	config,
 	fragmentGroupId,
 	fragmentKey,
+	network,
```
@julien we can do this and keep it as the previous implementation, but we will still need a dispatch for the NetworkService context, anyway we will still need to manually import the context or maybe we can make an observer to communicate directly with NetworkProvider, so we won't need to manually pass the context to all services. I will test this.

matuzalemsteles commented on modules/apps/layout/layout-content-page-editor-web/src/main/resources/META-INF/resources/page_editor/app/services/serviceFetch.js:
```diff
@@ -38,72 +37,63 @@ function getFormData({portletNamespace}, body) {
-						updateService({
+		return fetch(url, {
+			body: getFormData(config, body),
+			method: 'POST'
```
Sure, I think it's a good idea.

matuzalemsteles commented on modules/apps/layout/layout-content-page-editor-web/src/main/resources/META-INF/resources/page_editor/app/services/FragmentCommentService.js:
```diff
@@ -30,19 +28,21 @@ export default {
 	 * Adds a new Fragment to the current layout
 	 * @param {object} options
 	 * @param {string} options.body Body of the comment
+	 * @param {Function} options.fetcher service fetcher
```
No problem, I will test another strategy that I commented in another comment.

matuzalemsteles commented:
Resent #208
